### ISSUE TITLE
Improve GitHub Actions Hygiene

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2969]
GitHub Actions have recently been stagnating for days, causing significant delays in being able to review and merge PRs. While there have been a few issues on the administrative side, it looks like our actions are also consuming quota quite aggressively, and we should try to reduce that footprint, consolidating workflows and running them more selectively where possible.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* All jobs that do not depend on other jobs should be updated to run the `styfle/cancel-workflow-action` action as its first step
    * The dedicated `cancel.yml` workflow should be removed
* All workflows should define [`defaults`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#defaults) that sets [`timeout-minutes`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes) to `20` as a safeguard against run-away steps/jobs
* All workflows should define [`paths` and/or `paths-ignore`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths) to scope the run as a whole to only run when specified paths change
    *  Note: this is specific to the `push` or `pull_request` event configuration

## Implementation Brief

* Add `timeout-minutes: 20` as a default for all actions.
* Add `styfle/cancel-workflow-action` action to all actions as the first step aside from the `zips.yml` `deploy-to-wiki` job.
* Remove the `cancel.yml` action/file.
* Add `paths` for Jest/CSS/JS/PHP files accordingly, so JS tests are only run when JS files are changed, etc. (For now, always run E2E tests if any file other than a test file has changed, so that adding JS tests doesn't result in E2E tests being run…)
See #3046 for a PR that encompasses all changes sans the last one about combining jobs.

### Test Coverage

* N/A.

### Visual Regression Changes

* N/A.

## QA Brief

* GitHub Actions should only be run when code is pushed to a PR that affects said action—PHP changes should trigger PHPCS and PHP Unit runs, but not JS tests/linting, etc.
* Push a test branch to GitHub with only PHP changes. Observe that only PHP and E2E tests run, not JS/CSS-related tests.
* Push another commit with only JS/CSS changes. Observe only JS/CSS/E2E tests are run.

## Changelog entry

* N/A

**Comment by felixarntz** [comment]
@swissspidy What did you do for Web Stories to resolve these issues? Can you maybe share a high-level list or a link to issues/PRs?

**Comment by aaemnnosttv** [comment]
@felixarntz in the description you mention consolidating workflows, but the limit is on [concurrent jobs](https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits), not workflows. That is, we could have all jobs defined in a single workflow and it would be no different in terms of the limitations. That's good news because there are some things which can only be done with separate workflows, such as restarting or controlling when it runs.

We could however restructure some of our jobs a bit by consolidating workflows in a way which is more efficient in that we could then share the build between jobs (rather than running it separately in each) which is only possible between jobs of the same workflow. We could also maybe limit the number of jobs which run/queue right away by making some jobs depend on others. E.g. PHP Tests could depend on PHPCS to finish first. This would also allow for the Cancel action for subsequent runs to be triggered sooner (although I think there's more room for improvement here).

Another thing we should do is be more selective with [paths/files that trigger a workflow](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths). We should only run PHP tests when respective PHP classes change. One exception to this might be the tests against "WP latest/nightly" where something could break due to an upstream dependency change and we'd want to know about that regardless of whether or not our code changed.

**Comment by swissspidy** [comment]
Let me give you some context.

A couple of weeks ago a change was made that reduced the various limits for GitHub Actions for all repositories in this org, which led to all these issues with the queues piling up, nothing running for hours, etc. The change was reverted last Friday _for the time being_.

So for the time being, there's the expected [limit of 180 concurrent jobs](https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits) _for the whole organization_. Similar quotas exist for [minute and storage usage](https://docs.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions).

During all of this, google/site-kit and google/web-stories-wp were identified as two top consumers of GitHub Actions, scheduling tens of dozens of jobs, effectively blocking everyone else. At its peak google/site-kit-wp alone was at 140 concurrent jobs.

To help be better citizens, I've consulted with folks internally and implemented various changes to google/web-stories-wp to reduce GitHub Actions hours and consumption.

Here's the gist of it:

* Specify [`jobs.<job_id>.timeout-minutes`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes) for all jobs  
  Reduced maximum number of minutes to let a job run before GitHub automatically cancels it from 6 hours to 60 minutes.
* Removed the dedicated "Cancel previous" workflow in favor of adding the cancellation to every single job.  
  The problem with running it as a separate workflow is that said workflow has to run to cancel. In the event of a very large queue, it will never run, creating  a thundering herd problem
* Consolidated more tasks into the same workflow and/or job (https://github.com/google/web-stories-wp/pull/6733)  
  For example, while we had separate workflow files for JS/CSS/MD linting, they are now all in one job, reducing overhead (no need to spin up three runners)
* Set `paths`/`paths-ignore` for all relevant workflows (https://github.com/google/web-stories-wp/pull/6751)  
  This was quite drastic and still somewhat experimental.
  Pros: no unnecessary runs piling up, faster feedback for devs working on PRs.
  Cons: not possible to mark workflows as required status checks (as they now now longer always run). This puts some more responsibility/trust on folks with merge access, but it's manageable.  
  
  It's still a rather new change for us, so we'll have to see how well it works long term. But so far it's smooth.

**Comment by felixarntz** [comment]
@aaemnnosttv Can we please prioritize this? Would be great to add it to the upcoming sprint. cc @eclarke1 @fhollis

**Comment by felixarntz** [comment]
Can we potentially add the "Cancel previous" change to the ACs, so that it is included in every workflow instead being a separate one? I agree that makes sense and should be trivial.

Regarding use of a matrix, that's a bit more involved - also I'm not sure how a matrix would help. @swissspidy Does GH Actions count 3 different PHP test workflows vs 1 PHP test workflow with a matrix and 3 scenarios differently?

**Comment by swissspidy** [comment]
> Regarding use of a matrix, that's a bit more involved - also I'm not sure how a matrix would help. @swissspidy Does GH Actions count 3 different PHP test workflows vs 1 PHP test workflow with a matrix and 3 scenarios differently?

You'll only have a single run for a single job in the queue rather than 3 different ones. It will also reduce the total running time due to reduced overhead for setting up the runner and all the job's steps. And minutes spent also count towards [billing](https://docs.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions#about-billing-for-github-actions).

**Comment by felixarntz** [comment]
Thanks @swissspidy! @aaemnnosttv Maybe let's open a separate issue for ^ since it's a bit more involved of a change.

**Comment by aaemnnosttv** [comment]
> Can we potentially add the "Cancel previous" change to the ACs, so that it is included in every workflow instead being a separate one?

Are we sure that this is more effective? Either way it requires a newer job to start before a previous has finished for the previous to be cancelled (which isn't great) but it seems like the single job cancelling many at once would be more efficient in this regard, and this one is always very quick to run. The cancel action currently only cancels [workflow runs that were created before its own](https://github.com/styfle/cancel-workflow-action/blob/3d86a7cc43670094ac248017207be0295edbc31d/src/index.ts#L69) – ideally it would cancel all except the latest one (there's a PR or two that have attempted this). Without being able to cancel newer runs than itself, I think there will always be the potential for this to be a bottleneck that results in the queue filling up 😕 

Of course I'm not opposed to making this change if it works better than the dedicated workflow, I just don't see how it would be better than what we have now but maybe I'm not understanding it properly.  @felixarntz @swissspidy 

---

Other than this, is there anything else you'd like to see added here @felixarntz ? We also talked about consolidating a few jobs into a single workflow with dependencies so they run in sequence rather than in parallel, but I think we decided to leave this for a follow-up issue, correct?

**Comment by swissspidy** [comment]
> Are we sure that this is more effective? Either way it requires a newer job to start before a previous has finished for the previous to be cancelled (which isn't great) but it seems like the single job cancelling many at once would be more efficient in this regard, and this one is always very quick to run.

It doesn't matter how quick it is, if there are 50 other items  ahead of it in the queue.

In our experience so far this change made things work really well. To be really sure we could of course just do both.

**Comment by felixarntz** [comment]
@aaemnnosttv 
> Other than this, is there anything else you'd like to see added here @felixarntz ? We also talked about consolidating a few jobs into a single workflow with dependencies so they run in sequence rather than in parallel, but I think we decided to leave this for a follow-up issue, correct?

Yeah, I think any rearrangement of jobs (other than the "Cancel" one) should be left for a follow-up issue.

**Comment by aaemnnosttv** [comment]
IB ✅ 

> I think there's more we could do here like combining JS Tests and JS/CSS Linting jobs into one; it would be straightforward to do in this PR without getting into things like `matrix:` jobs, and would reduce our simultaneous jobs with little effort.

Let's keep the ACs as-is for now and we can address this in a new issue as an immediate follow-up.

**Comment by ivankruchkoff** [comment]
QA ✅

Created two test branches `test/2969-js` and `test/2969-php` and draft PRs for them. Put a js dummy commit in the js branch and php dummy commit in the php branch, and looked through the actions that were run.

![screencapture-github-google-site-kit-wp-actions-2021-04-07-11_10_41](https://user-images.githubusercontent.com/373931/113919515-8a892900-9798-11eb-9644-06429fd26502.png)
